language/internal/cc/parser/parser.go

```diff
@@ -107,37 +137,470 @@ func tokenizer(data []byte, atEOF bool) (advance int, token []byte, err error) {
 	return i, nil, nil
 }
 
-func extractSourceInfo(input io.Reader) SourceInfo {
-	scanner := bufio.NewScanner(input)
-	scanner.Split(tokenizer)
+type parser struct {
+	tr        *tokenReader
+	lastToken string
+
+	// accumulated result
+	sourceInfo SourceInfo
+
+	// active #if/#else nesting – conjunction of these is the current guard
+	conditionStack []Expr
+	// stack of "already‑seen" branch expressions for each #if group;
+	// used to build !previous when we see #else / #elif
+	exprGroupStack [][]Expr
```

Comment:
Optional: I found this evaluation model a little confusing.

As I understand, we're parsing the file once, producing a list of includes, and given each include a conditional expression. That works, but it's hard for me to tell if the expressions are going to be O(n^2) in the size of the program. We can certainly evaluate the sub-expressions more than once when a group of includes is covered by the same expressions.

I'd suggest instead:

- Incorporate the includes into the syntax tree. An individual include won't have a condition: the condition is implied by its place in the tree.
- Later on, we'll have a function that evaluates the tree given an environment (a map from identifiers to integers). It's essentially a tree-walking interpreter producing a list of includes.